lib/pure/osproc.nim

```diff
@@ -434,6 +434,87 @@ when not defined(useNimRtl):
 template streamAccess(p) =
   assert poParentStreams notin p.options, "API usage error: stream access not allowed when you use poParentStreams"
 
+# Wrap pipe for reading with PipeOutStream so that you can use peek* and generate runtime error
+# when setPosition/getPosition is called or write operation is performed.
+type
+  PipeOutStream[T] = ref object of T
+    buffer: Deque[char]
+    baseReadLineImpl: typeof(StreamObj.readLineImpl)
+    baseReadDataImpl: typeof(StreamObj.readDataImpl)
+
+proc posReadLine[T](s: Stream, line: var TaintedString): bool =
+  var s = PipeOutStream[T](s)
+  assert s.baseReadLineImpl != nil
+
+  let n = s.buffer.len
+  line.string.setLen(0)
+  for i in 0..<n:
+    var c = s.buffer.popFirst
+    if c == '\c':
+      c = readChar(s)
+      return true
+    elif c == '\L': return true
+    elif c == '\0':
+      return line.len > 0
+    line.string.add(c)
+
+  var line2: string
+  result = s.baseReadLineImpl(s, line)
+  line.add line2
+
+proc posReadData[T](s: Stream, buffer: pointer, bufLen: int): int =
+  var s = PipeOutStream[T](s)
+  assert s.baseReadDataImpl != nil
+
+  let
+    dest = cast[ptr UncheckedArray[char]](buffer)
+    n = min(s.buffer.len, bufLen)
+  result = n
+  for i in 0..<n:
+    dest[i] = s.buffer.popFirst
+  if bufLen > n:
+    result += s.baseReadDataImpl(s, addr dest[n], bufLen - n)
+
+proc posReadDataStr[T](s: Stream, buffer: var string, slice: Slice[int]): int =
+  posReadData[T](s, addr buffer[slice.a], slice.len)
+
+proc posPeekData[T](s: Stream, buffer: pointer, bufLen: int): int =
+  var s = PipeOutStream[T](s)
+  assert s.baseReadDataImpl != nil
+
+  let n = min(s.buffer.len, bufLen)
+  result = n
+  if bufLen > n:
+    let m = bufLen - n
+    var buf = newSeq[char](m)
```

Comment:
Rather than create a sequence here, is it possible to extend the existing buffer, and read data directly into it?
Something like:
```
  let dest = cast[ptr UncheckedArray[char]](buffer)

  let newDataNeeded = bufLen - s.buffer.len
  if newDataNeeded > 0:
    # We need more data than the internal buffer can supply, so
    # use the user's buffer to retrieve the new data
    result = bufLen
    result += s.baseReadDataImpl(s, addr dest[s.buffer.len], newDataNeeded)

    # Copy the user buffer's new data into the stream buffer
    for i in 0 ..< newDataNeeded:
      s.buffer.addLast dest[i + s.buffer.len]
    # Copy the stream buffer's old data into the user buffer
    for i in 0 .. s.buffer.len:
      dest[i] = s.buffer[i]

  else:
    # Copy some of the stream buffer data into the user buffer
    for i in 0 ..< bufLen:
      dest[i] = s.buffer[i]
```